validator/client/metrics.go

```diff
@@ -161,66 +161,110 @@ func (v *validator) LogValidatorGainsAndLosses(ctx context.Context, slot uint64)
 			}).Info("Previous epoch voting summary")
 			if v.emitAccountMetrics {
 				ValidatorBalancesGaugeVec.WithLabelValues(fmtKey).Set(newBalance)
 			}
 		}
+		v.prevBalance[pubKeyBytes] = resp.BalancesBeforeEpochTransition[i]
+	}
+
+	v.UpdateLogAggregateStats(resp, slot)
+	return nil
+}
 
+//UpdateLogAggregateStats updates and logs the summaryStats struct of a validator using the RPC response obtained during LogValidatorGainsAndLosses
+func (v *validator) UpdateLogAggregateStats(resp *ethpb.ValidatorPerformanceResponse, slot uint64) {
+	summary := &v.summaryStats
+	currentEpoch := slot / params.BeaconConfig().SlotsPerEpoch
+	var included, correctSource, correctTarget, correctHead int
+
+	for i := range resp.PublicKeys {
 		if resp.InclusionSlots[i] != ^uint64(0) {
 			included++
+			summary.includedAttests++
+			summary.totalDistance += resp.InclusionDistances[i]
 		}
 		if resp.CorrectlyVotedSource[i] {
-			votedSource++
+			correctSource++
+			summary.correctSources++
 		}
 		if resp.CorrectlyVotedTarget[i] {
-			votedTarget++
+			correctTarget++
+			summary.correctTargets++
 		}
 		if resp.CorrectlyVotedHead[i] {
-			votedHead++
+			correctHead++
+			summary.correctHeads++
 		}
-		v.prevBalance[pubKeyBytes] = resp.BalancesBeforeEpochTransition[i]
 	}
 
+	summary.totalAttestSlots += uint64(len(resp.InclusionSlots))
+	summary.totalSources += uint64(len(resp.CorrectlyVotedSource))
+	summary.totalTargets += uint64(len(resp.CorrectlyVotedTarget))
+	summary.totalHeads += uint64(len(resp.CorrectlyVotedHead))
+
 	log.WithFields(logrus.Fields{
-		"epoch":                          prevEpoch,
-		"attestationInclusionPercentage": fmt.Sprintf("%.0f%%", (float64(included)/float64(len(resp.InclusionSlots)))*100),
-		"correctlyVotedSourcePercentage": fmt.Sprintf("%.0f%%", (float64(votedSource)/float64(len(resp.CorrectlyVotedSource)))*100),
-		"correctlyVotedTargetPercentage": fmt.Sprintf("%.0f%%", (float64(votedTarget)/float64(len(resp.CorrectlyVotedTarget)))*100),
-		"correctlyVotedHeadPercentage":   fmt.Sprintf("%.0f%%", (float64(votedHead)/float64(len(resp.CorrectlyVotedHead)))*100),
+		"epoch":                   currentEpoch - 1,
+		"attestationInclusionPct": fmt.Sprintf("%.0f%%", (float64(included)/float64(len(resp.InclusionSlots)))*100),
+		"correctlyVotedSourcePct": fmt.Sprintf("%.0f%%", (float64(correctSource)/float64(len(resp.CorrectlyVotedSource)))*100),
+		"correctlyVotedTargetPct": fmt.Sprintf("%.0f%%", (float64(correctTarget)/float64(len(resp.CorrectlyVotedTarget)))*100),
+		"correctlyVotedHeadPct":   fmt.Sprintf("%.0f%%", (float64(correctHead)/float64(len(resp.CorrectlyVotedHead)))*100),
 	}).Info("Previous epoch aggregated voting summary")
 
-	return nil
+	var totalStartBal, totalPrevBal uint64
+	for i, val := range v.startBalance {
+		totalStartBal += val
+		totalPrevBal += v.prevBalance[i]
+	}
+
+	log.WithFields(logrus.Fields{
+		//+1 to adjust for the original startEpoch calculation
```

Comment:
Not sure what this comment means. Can you clarify?

Reply:
Yes I was having some issues with the wording.  In metrics.go function LogValidatorGainsAndLosses about line 165 there is this piece of code.  

     if slot >= params.BeaconConfig().SlotsPerEpoch {
		prevEpoch = (slot / params.BeaconConfig().SlotsPerEpoch) - 1
		if v.voteStats.startEpoch == 0 {
			//prevEpoch may be 0 so we store startEpoch (+1). Thus a value of 0 tells us it is not set (default)
			v.voteStats.startEpoch = prevEpoch + 1
		}
	}

Okay:  startEpoch = 0 (default value) after initialization (as it is untouched) .  It is then set during the first log report of this kind -- where prevEpoch is obtained.  Now the function is only entered after epoch 0 finishes (it returns early if it is within the 0th epoch) so the earliest epoch is epoch 1 where this logic is reached.  prevEpoch may equal 0  (1 - 1).  Essentially I wanted a way to tell if the epoch at validator start was the default value (unset) or not.  Thus I elected to store "startEpoch" instead of prevEpoch in the stats structure I created.  I.e. startEpoch can never be calculated to be 0, thus 0 must mean uninitialized.  Anyways that is just the backstory.

Anyways, when we output statistics, it is up to the last epoch.  So say the validator started during epoch 10 start.  (prev)Epoch 9 is logged out.  startEpoch is set to 10 at that time.  currentEpoch is 10.  So we if we currentEpoch - startEpoch we get 0 even though we have logged out 1 epoch in total.  The +1 adjusts for that.  We could do currentEpoch - (startEpoch- 1) which would be the same as currentEpoch - startEpoch +1.

How can I improve the comment?  I wanted to explain the +1 but had a hard time writing it down.  
`//+1 to include the original epoch`
?

Reply:
Taken care of in a recent commit addressing subsequent feedback.